neqo-transport/src/connection/mod.rs

```diff
@@ -2753,32 +2753,9 @@ impl Connection {
 
     fn cleanup_streams(&mut self) {
         self.send_streams.clear_terminal();
-        let recv_to_remove = self
-            .recv_streams
-            .iter()
-            .filter_map(|(id, stream)| {
-                // Remove all streams for which the receiving is done (or aborted).
-                // But only if they are unidirectional, or we have finished sending.
-                if stream.is_terminal() && (id.is_uni() || !self.send_streams.exists(*id)) {
-                    Some(*id)
-                } else {
-                    None
-                }
-            })
-            .collect::<Vec<_>>();
-
-        let mut removed_bidi = 0;
-        let mut removed_uni = 0;
-        for id in &recv_to_remove {
-            self.recv_streams.remove(&id);
-            if id.is_remote_initiated(self.role()) {
-                if id.is_bidi() {
-                    removed_bidi += 1;
-                } else {
-                    removed_uni += 1;
-                }
-            }
-        }
+        let send_streams = &self.send_streams;
+        let (removed_bidi, removed_uni) =
+            self.recv_streams.clear_terminal(send_streams, self.role());
```

Comment:
I am not really happy about this. I decided to refactor it when I remove the last part of FlowMgr (the flow control for open streams)